cli/src/datafeed.ts

```diff
@@ -112,6 +145,73 @@ export function bulkInsertHandler() {
   return wrapper
 }
 
+export function bulkRemoveHandler() {
+  async function wrapper({ data }: { data }) {
+    const bulkData = data as IDatafeedBulk
+
+    if (!checkBulkSource(data?.bulk)) {
+      console.error('invalid json src format')
+      return
+    }
+
+    const adapters = await adapterListHandler()()
+    const aggregators = await aggregatorListHandler()({})
+    const listeners = await listenerListHandler()({})
+    const reporters = await reporterListHandler()({})
+    const delegatorReporters = await delegatorReporterListHandler()()
+    const delegatorContracts = await contractListHandler()()
+    const delegatorFunctions = await functionListHandler()()
+
+    for (const removeElement of bulkData.bulk) {
+      console.log(`removing ${removeElement}`)
+      const adapterData = await loadJsonFromUrl(removeElement.adapterSource)
+      if (!checkAdapterSource(adapterData)) {
+        console.error(`invalid adapterData: ${adapterData}, skipping insert`)
+        continue
+      }
+      const aggregatorData = await loadJsonFromUrl(removeElement.aggregatorSource)
+      if (!checkAggregatorSource(aggregatorData)) {
+        console.error(`invalid aggregatorData: ${aggregatorData}, skipping insert`)
+        continue
+      }
+
+      const adapterId = adapters.find(
+        (adapter) => adapter.adapterHash == adapterData.adapterHash
+      ).id
+      const aggregatorId = aggregators.find(
+        (aggregator) => aggregator.aggregatorHash == aggregatorData.aggregatorHash
+      ).id
+
+      const listenerId = listeners.find((listener) => listener.address == aggregatorData.address).id
+      const reporterId = reporters.find(
+        (reporter) => reporter.address == removeElement.reporter.walletAddress
+      ).id
+
+      const delegatorReporterId = delegatorReporters.find(
+        (reporter) =>
+          reporter.address.toLowerCase() == removeElement.reporter.walletAddress.toLowerCase()
+      ).id
+      const delegatorContractId = delegatorContracts.find(
+        (contract) => contract.address.toLowerCase() == aggregatorData.address.toLowerCase()
+      ).id
+      const functionId = delegatorFunctions.find(
+        (_function) => _function.address.toLowerCase() == aggregatorData.address.toLowerCase()
+      ).id
+
+      await listenerRemoveHandler()({ id: listenerId })
+      await reporterRemoveHandler()({ id: reporterId })
+
+      await functionRemoveHandler()({ id: functionId })
+      await contractRemoveHandler()({ id: delegatorContractId })
+      await delegatorReporterRemoveHandler()({ id: delegatorReporterId })
+
+      await aggregatorRemoveHandler()({ id: aggregatorId })
```

Comment:
I think in the real system, it would not be possible to remove `aggregator` and `adapter` if they have already collected some data. They are linked through foreign keys to `data` table which does not allow you to delete them without deleting data at first. I think we should avoid deleting data, and it does not hurt to keep old adapter and aggregator. Their unique hashes are computed from their definitions which would not make issues when we add newer version of the same data feed. How about not deleting them in here? 🤔